---
author-meta:
- Sharon Waymost
- Serghei Mangul
- Neha Rajkumar
- Ram Ayyala
- Heng Li
- Nathan LaPierre
- "Andr\xE9 Santos"
- Titus Brown
- Casey S. Greene
- Brendan Lawlor
- Qiyang Hu
- Jaqueline J. Brito
- Thiago Mosqueiro
bibliography:
- content/manual-references.json
date-meta: '2020-09-03'
header-includes: "<!--\nManubot generated metadata rendered from header-includes-template.html.\nSuggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html\n-->\n<meta name=\"dc.format\" content=\"text/html\" />\n<meta name=\"dc.title\" content=\"Packaging and containerizing of bioinformatics software: advances, challenges, and opportunities\" />\n<meta name=\"citation_title\" content=\"Packaging and containerizing of bioinformatics software: advances, challenges, and opportunities\" />\n<meta property=\"og:title\" content=\"Packaging and containerizing of bioinformatics software: advances, challenges, and opportunities\" />\n<meta property=\"twitter:title\" content=\"Packaging and containerizing of bioinformatics software: advances, challenges, and opportunities\" />\n<meta name=\"dc.date\" content=\"2020-09-03\" />\n<meta name=\"citation_publication_date\" content=\"2020-09-03\" />\n<meta name=\"dc.language\" content=\"en-US\" />\n<meta name=\"citation_language\" content=\"en-US\" />\n<meta name=\"dc.relation.ispartof\" content=\"Manubot\" />\n<meta name=\"dc.publisher\" content=\"Manubot\" />\n<meta name=\"citation_journal_title\" content=\"Manubot\" />\n<meta name=\"citation_technical_report_institution\" content=\"Manubot\" />\n<meta name=\"citation_author\" content=\"Sharon Waymost\" />\n<meta name=\"citation_author_institution\" content=\"Computer Science Department, Samueli School of Engineering, University of California, Los Angeles\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-1176-5386\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Serghei Mangul\" />\n<meta name=\"citation_author_institution\" content=\"Department of Clinical Pharmacy, School of Pharmacy, University of Southern California\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-4770-3443\" />\n<meta name=\"citation_author\" content=\"Neha Rajkumar\" />\n<meta name=\"citation_author_institution\" content=\"Department of Bioengineering, Samueli School of Engineering, University of California, Los Angeles\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Ram Ayyala\" />\n<meta name=\"citation_author_institution\" content=\"Department of Neuroscience, School of Life Sciences, University of California Los Angeles\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Heng Li\" />\n<meta name=\"citation_author_institution\" content=\"None\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Nathan LaPierre\" />\n<meta name=\"citation_author_institution\" content=\"None\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Andr\xE9 Santos\" />\n<meta name=\"citation_author_institution\" content=\"None\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Titus Brown\" />\n<meta name=\"citation_author_institution\" content=\"University of California, Davis\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Casey S. Greene\" />\n<meta name=\"citation_author_institution\" content=\"Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine, University of Pennsylvania\" />\n<meta name=\"citation_author_institution\" content=\"Childhood Cancer Data Lab\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-8713-9213\" />\n<meta name=\"citation_author\" content=\"Brendan Lawlor\" />\n<meta name=\"citation_author_institution\" content=\"Department of Computer Science, Cork Institute of Technology, Cork, Ireland.\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-2250-0669\" />\n<meta name=\"twitter:creator\" content=\"@blawlor\" />\n<meta name=\"citation_author\" content=\"Qiyang Hu\" />\n<meta name=\"citation_author_institution\" content=\"None\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<meta name=\"citation_author\" content=\"Jaqueline J. Brito\" />\n<meta name=\"citation_author_institution\" content=\"Department of Clinical Pharmacy, School of Pharmacy, University of Southern California\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-7158-3253\" />\n<meta name=\"twitter:creator\" content=\"@jaquejbrito\" />\n<meta name=\"citation_author\" content=\"Thiago Mosqueiro\" />\n<meta name=\"citation_author_institution\" content=\"None\" />\n<meta name=\"citation_author_orcid\" content=\"None\" />\n<meta name=\"twitter:creator\" content=\"@None\" />\n<link rel=\"canonical\" href=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/\" />\n<meta property=\"og:url\" content=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/\" />\n<meta property=\"twitter:url\" content=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/\" />\n<meta name=\"citation_fulltext_html_url\" content=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/\" />\n<meta name=\"citation_pdf_url\" content=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"application/pdf\" href=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"text/html\" href=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/v/62f4cc2a01bb76f8f207849769c05d900e88d655/\" />\n<meta name=\"manubot_html_url_versioned\" content=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/v/62f4cc2a01bb76f8f207849769c05d900e88d655/\" />\n<meta name=\"manubot_pdf_url_versioned\" content=\"https://Mangul-Lab-USC.github.io/pkg-manager-review/v/62f4cc2a01bb76f8f207849769c05d900e88d655/manuscript.pdf\" />\n<meta property=\"og:type\" content=\"article\" />\n<meta property=\"twitter:card\" content=\"summary_large_image\" />\n<link rel=\"icon\" type=\"image/png\" sizes=\"192x192\" href=\"https://manubot.org/favicon-192x192.png\" />\n<link rel=\"mask-icon\" href=\"https://manubot.org/safari-pinned-tab.svg\" color=\"#ad1457\" />\n<meta name=\"theme-color\" content=\"#ad1457\" />\n<!-- end Manubot generated metadata -->"
keywords:
- package managers
- containerization
- distribution
- docker
lang: en-US
manubot-clear-requests-cache: false
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
title: 'Packaging and containerizing of bioinformatics software: advances, challenges, and opportunities'
...






<small><em>
This manuscript
([permalink](https://Mangul-Lab-USC.github.io/pkg-manager-review/v/62f4cc2a01bb76f8f207849769c05d900e88d655/))
was automatically generated
from [Mangul-Lab-USC/pkg-manager-review@62f4cc2](https://github.com/Mangul-Lab-USC/pkg-manager-review/tree/62f4cc2a01bb76f8f207849769c05d900e88d655)
on September 3, 2020.
</em></small>

## Unsorted list of authors



+ **Sharon Waymost**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-1176-5386](https://orcid.org/0000-0003-1176-5386)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [sbpw](https://github.com/sbpw)<br>
  <small>
     Computer Science Department, Samueli School of Engineering, University of California, Los Angeles
  </small>

+ **Serghei Mangul**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-4770-3443](https://orcid.org/0000-0003-4770-3443)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [smangul1](https://github.com/smangul1)<br>
  <small>
     Department of Clinical Pharmacy, School of Pharmacy, University of Southern California
  </small>

+ **Neha Rajkumar**<br><br>
  <small>
     Department of Bioengineering, Samueli School of Engineering, University of California, Los Angeles
  </small>

+ **Ram Ayyala**<br><br>
  <small>
     Department of Neuroscience, School of Life Sciences, University of California Los Angeles
  </small>

+ **Heng Li**<br><br>
  <small>
     None
  </small>

+ **Nathan LaPierre**<br><br>
  <small>
     None
  </small>

+ **André Santos**<br><br>
  <small>
     None
  </small>

+ **Titus Brown**<br><br>
  <small>
     University of California, Davis
  </small>

+ **Casey S. Greene**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-8713-9213](https://orcid.org/0000-0001-8713-9213)<br>
  <small>
     Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine, University of Pennsylvania; Childhood Cancer Data Lab
  </small>

+ **Brendan Lawlor**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-2250-0669](https://orcid.org/0000-0002-2250-0669)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [blawlor](https://github.com/blawlor)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [blawlor](https://twitter.com/blawlor)<br>
  <small>
     Department of Computer Science, Cork Institute of Technology, Cork, Ireland.
  </small>

+ **Qiyang Hu**<br><br>
  <small>
     None
  </small>

+ **Jaqueline J. Brito**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-7158-3253](https://orcid.org/0000-0002-7158-3253)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jaquejbrito](https://github.com/jaquejbrito)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [jaquejbrito](https://twitter.com/jaquejbrito)<br>
  <small>
     Department of Clinical Pharmacy, School of Pharmacy, University of Southern California
  </small>

+ **Thiago Mosqueiro**<br><br>
  <small>
     None
  </small>



## Abstract {.page_break_before}




## Introduction {.page_break_before}

The rapid advancement of omics and sequencing technologies has led to an accelerated growth of genomic data.
Driven by this data, there are similar increases in the number of available bioinformatics software tools, often containing novel and diverse algorithms of particular interest to biomedical researchers.[1,2].
Biomedical researchers wanting to use this software often have access to high-performance clusters (“clusters”), but lack the operating system-level permissions and advanced skills required to install and run the software.
None of the necessary skills are currently included in the traditional life science curriculum at major universities.
Usability is further hindered by the absence of any standardization and the wide variety of software development tools employed.
This is directly observable in many academic software tools lacking a user-friendly interface[5].

As the dependence of biomedical researchers on computational software continues to increase, software developers need to consider more user-friendly distribution and install methods.
Modern software distribution, having already faced user-friendliness issues, is already becoming more reliant on platforms such as package managers[6,7] and containers[8].
Both promise to simplify software development while increasing the usability and reproducibility of biomedical research[8,9].

Package managers first appeared nearly thirty years ago as software developers sought to streamline the entire installation process.
To install via package manager, the user must only specify the desired software, called a “package;” the download, installation, configuration, and dependencies are all handled by the package manager. 
Many operating systems have built-in package managers (e.g. APT), which may not be available to cluster users, while others must be downloaded and installed by the user. 
Some package managers (e.g. Conda[6]) are programming language agnostic, while others are designed for a particular language (e.g pip[10]).

A more recent software distribution solution is containerization.
The end-user downloads a container “image” that includes the software, dependencies, and anything else necessary to run the software.
Though the imaged software is not typically itself installed, many do require the installation of containerization runtime software. 
When run, the runtime software creates a consistent, isolated sandbox environment, then runs the imaged software inside the sandbox[11].
This sandbox design makes the images both highly portable (compatible with different computers) and easily shareable (transferable between different computers), which has already led to wide adaptation in bioinformatics[7,12].

In addition to the ease of installation and use offered by both package managers and containers, such platforms must also meet the biomedical community’s need for compatibility with high performance clusters.
However, the  relative performance of these package managers and containers remains unknown.
Our review summarizes developing practices across the most common package managers and containers, while discussing the challenges, advantages, and limitations of using them from both the developer and the user perspectives.
By taking a survey of all the available package managers and container software, there is now a comprehensive list of the different attributes of each one, informing the community so that people can make educated decisions about which package manager or container makes the most sense for their project.
We also propose principles that can make packaging and containerizing of bioinformatics software more sustainable and reproducible, ultimately increasing the usability of bioinformatics software.


## Discussion {.page_break_before}

### Existing problems with software distribution and installation

The installation process of bioinformatics research software is typically a multi-step process, starting with the end-user locating and downloading the software. 
Next is the actual installation, during which the end-user must determine what dependencies are missing and resolve them by installing the required software[5](https://paperpile.com/c/4vBDtY/kjwlC).
Even in cases where the end-user is familiar with this process, they are typically installing on high-performance clusters where they are constrained by user-level permissions that prevent them from following standard installation procedures.



- root access limitations
- reproducibility of findings
- version conflicts
- dependency resolution

### Definitions and explanations of distribution system types

#### Package Managers

Package managers appeared nearly thirty years ago as software developers sought to automate and standardize the software installation process.
The earliest package managers (e.g. APT and RPM) were built into Linux operating systems and lacked now common features like dependency resolution.
Modern package managers are available for all major operating systems, with both integrated and user-installed options available.
Some are package managers are limited to software written in a specific programming language (e.g. pip distributes software written in Perl), while others are limited to sofware with a particular purpose (e.g. Bioconda distributes omics software).

Every package managers uses its own package format, which follows the general structure shown in Figure @fig:package-schematic. 

![
**Schematic showing standard package contents.**
Though the exact contents varies between different package formats, all include the software and relevant metadata necessary to complete installation.
](images/package_schematic.png){#fig:package-schematic}

The metadata gives the package manager the necessary information to install the packaged software. 

To use a package manager, a user instructs the package manager to install software, then the package manager performs all the necessary steps to install the software as shown in Figure @fig:package-manager-usage.
Assuming the software is not already installed, the package manager retrieves the appropriate package from the repository, a server that stores all of the available packages.
The package manager then uses the package's metadata to determine what it needs to do to complete installation.
For most packages, this includes verifying the installation of all dependencies, which are other pieces of software that must be installed prior to the initial software being installed.

![
**Standard workflow for installing software via a package manager.**
The user, usually an admin, asks the package manager to install a specific piece of software. If the software is not already installed, the package manager fetches the approrpriate package from the repository. If any of the dependencies are not already installed, the package manager "loops" through the missing dependencies. Once all dependencies are installed, the original software is installed.  
](images/package_manager_usage.png){#fig:package-manager-usage width="6in"}

The package manager verifies each dependency by confirming that the dependency is already installed or installing the dependency if it is missing. 
For missing dependencies, the package manager retrieves the dependency's package from the repository and starts the installation procedure for that package.
Once all the dependencies are installed, the initially requested software is installed. 
However, the package manager often goes through several iterations of this process because every dependency can have it's own list of dependencies, in which case each of the dependency's dependencies must be verified through the same process.

In most cases, a package manager is able to install all of the missing dependencies, but there are two relatively common problems that a package manager cannot overcome.
The first is circular dependencies, which occurs when software A has a dependency on software B and software B has a dependency on software A.
A package manager cannot install A without installing B first, but B cannot be installed without installing A first.
The second is conflicting versions, which occurs when software A and software B have the same dependency, but require different versions.
Package managers only allow one version of software to be installed at a time, preventing both A and B from having their depedencies met simultaneously.

Administrators often block users from installing software to prevent unwanted changes to both the operating system and existing sofware. This includes blocking access to package managers built into the operating system, as well as preventing users from installing their own.


  - benefits for the developer
    - mature technology - higher degree of familiarity
    - allows dependency specification (including versions)
    -limitations for the developer
    - can't always use to install missing dependencies for end-user
  - benefits for the end-user
    - package size is minimal (dependencies aren't duplicated)
    - installs missing dependencies
  - limitations for the end-user
    - not always accessible (unless admin user)
    - can't install multiple versions of same software
- containerization
  - definition

![
**Schematic showing standard containerization objects.**
](images/containerization_schematics.png){#fig:containerization-schematics width="6in"}

![
**Running containerized software**
](images/container_usage.png){#fig:container_usage width="6in"}

  - benefits for the developer
    - include specific versions of dependencies
    - known running environment
      - fewer test variables
      - reproducibility of results
  - limitations for the developer
    - learn a new system instead of focusing on research
  - benefits for the end-user
    - no installation (except possible runtime)
    - no dependency issues
    - sandbox provides computer system security
  - limitations for the end-user
    - container size
    - duplication of dependencies
    - root access requirement to install runtime
    - configuration in cluster
- centralized repositories
  - definition
    - channels
  - benefits
    - known download site
    - hosting is taken of
  - limitations
    - repo specific restrictions

![
**Relative popularity of package managers.**
According to Google Trends, general use package managers have at least ten times the popularity of their omics-only counterparts, which is expected when their limited-use is taken into account.
Docker, the most popular containerization software is nearly five times as popular as the most popular package manager, indicating a software industry in <location> trend towards containerization over the past <time period>.
](images/trends.png){#fig:google-trends-chart width="6in"}

![
**Usage of package managers in academia based on citations.**
](images/citations.png){#fig:citations-chart width="6in"}

![
**Usage of package managers in academia based on Google Scholar Results.**
](images/google_scholar_results.png){#fig:google-scholar-chart width="6in"}

![
**Total number of packages.**
](images/total_number_of_packages.png){#fig:available-packages-chart width="6in"}

![
**Packages heatmap**
](images/tools_heatmap.png){#fig:available-packages-heatmap width="6in"}

![
**Overlap in available packages between omics package managers.**
](images/overlap_in_tools.png){#fig:overlap-in-tools width="6in"}

![
**Availablity of Bioconda's packages in non-omics package managers.**
](images/tools_shared_with_bioconda.png){#fig:bioconda-packages-availability-chart width="6in"}

![
**Availablity of Bioinformatics Tools' packages in non-omics package managers.**
](images/tools_shared_with_biotools.png){#fig:biotools-packages-availability-chart width="6in"}

![
**Availability of omicX's packages in non-omics package managers.**
](images/tools_shared_with_omicX.png){#fig:omicX-packages-availability-chart width="6in"}


## Methods {.page_break_before}

**@fig:google-trends-chart**
Google Trends @https://trends.google.com/trends/ was used to assess the relative average popularity of each package manager. 
Package managers were omitted when Google Trends did not have enough data to return a result. 
The method detailed in "How do you compare large number of items in Google Trends?" @https://digitaljobstobedone.com/2017/07/10/how-do-you-compare-large-numbers-of-items-in-google-trends/ was used to overcome Google Trends' limitation of five search terms. 

@fig:citations-chart
For package managers introduced with a major publication, the number of citations for their publication was found via Google Scholar @https://scholar.google.com/ and recorded.
Package managers without a major publication were excluded.

@fig:google-scholar-chart
Each package manager's name was used as a search term on Google Scholar @https://scholar.google.com/ and the number of results was recorded.
Package manager names that returned results for things other than the package manager were omitted.

@fig:available-packages-chart
The initial list of all available sofware for each package manager was compiled by parsing its online repository with WebScraper @https://webscraper.io/
(AppImage @AppImage-packages,
Bioconda @Bioconda-packages,
Bioconductor @Bioconductor-packages,
BioContainers @Biocontainers-packages,
CRAN @CRAN-packages,
Docker @Docker-packages,
EasyBuild @EasyBuild-packages,
Flatpak @Flatpak-packages,
GNU Guix @GNU_Guix-packages,
omicX @omicX-packages,
pip @pip-packages,
Singularity @Singularity-packages,
Snap @Snap-packages,
Spack @Spack-packages,
Vagrant @Vagrant-packages,
and Zero Install @Zero_Install-packages)
or by querying the package manager directly (APT, Biocontainers, Conda, Homebrew).
Each list converted to a set with pandas @https://pandas.pydata.org/ and NumPy @https://numpy.org/, which removed all capitalization and duplicates.
The number of entries in a package manager's set was taken to be the total number of packages available for that package manager.

@fig:available-packages-heatmap
The sets obtained in @fig:available-packages-chart were compared with the set intersection function to find the number of commonalities between all pairs of package managers.

@fig:bio-package-managers-venn-diagram
The sets obtained for Bioconda, Bioconductor, and omicX in @fig:available-packages-chart were compared with the set intersection function to find the number of commonalities between all combinations of the three package managers.

@fig:bioconda-packages-availability-chart
The sets obtained for Bioinformatics Tools and the non-omics package managers in @fig:available-packages-chart were compared with the set intersection function to find the number of commonalities between Bioconda and the non-omics package managers.

@fig:biotools-packages-availability-chart
The sets obtained for Bioinformatics Tools and the non-omics package managers in @fig:available-packages-chart were compared with the set intersection function to find the number of commonalities between Bioinformatics Tools and the non-omics package managers.

@fig:omicX-packages-availability-chart
The sets obtained for omicX and the non-omics package managers in @fig:available-packages-chart were compared with the set intersection function to find the number of commonalities between omicX and the non-omics package managers.

## Glossary {.page_break_before}




## Acknowledgements {.page_break_before}




## Author Contributions {.page_break_before}




## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs">

<!-- References for HTML versions of repositories -->
@AppImage-packages: https://appimage.github.io/apps/
@Bioconda-packages: https://bioconda.github.io/conda-recipe_index.html
@Bioconductor-packages: http://bioconductor.org/packages/release/BiocViews.html#___Software
@BioContainers-packages: https://github.com/BioContainers/containers
@CRAN-packages: https://cran.r-project.org/web/packages/available_packages_by_name.html
@Docker-packages: https://hub.docker.com/search?q=&type=image&page=1
@EasyBuild-packages: https://easybuild.readthedocs.io/en/latest/version-specific/Supported_software.html
@Flatpak-packages: https://flathub.org/apps/category/All
@GNU_Guix-packages: https://guix.gnu.org/packages/ 
@omicX-packages: https://omictools.com/software?page=1
@pip-packages: https://pypi.org/simple/
@Singularity-packages: https://singularity-hub.org/apps
@Snap-packages: https://snapcraft.io/store
@Spack-packages: https://spack.readthedocs.io/en/latest/package_list.html
@Vagrant-packages: https://app.vagrantup.com/boxes/search?order=desc&page=2375
@Zero_Install-packages: http://roscidus.com/0mirror/feed-list


</div>


## Tables {.page_break_before}

| Distribution System Name | URL | Publication | Type | License |
| :----------- | :---------------------------------------- | :-------------------- | :--------------- | :------------ |
| AppImage     | https://appimage.org                      | -                     | containerization | MIT           |
| APT          | https://wiki.debian.org/Apt               | -                     | package manager  | GNU GPL 2+    |
| Bioconda     | https://bioconda.github.io                | Grüning et al, 2018   | package manager  | MIT           |
| Bioconductor | https://www.bioconductor.org              | Gentleman et al, 2004 | package manager  | MIT           |
| conda        | https://docs.conda.io/en/latest/          | -                     | package manager  | 3-Clause BSD  |
| CRAN         | https://cran.r-project.org/index.html     | -                     | package manager  | GNU GPL       |
| Docker       | https://www.docker.com                    | -                     | containerization | Apache 2.0    |
| Easybuild    | https://easybuilders.github.io/easybuild/ | Hoste et al, 2012     | package manager  | GNU GPL 2     |
| Flatpak      | https://flatpak.org                       | -                     | containerization | LGPL          |
| GNU Guix     | https://guix.gnu.org                      | Courtès, 2013         | package manager  | GNU AGPL      |
| Homebrew     | https://brew.sh                           | -                     | package manager  | 2-Clause BSD  |
| pip          | https://pypi.org/project/pip/             | -                     | package manager  | MIT           |
| Singularity  | https://sylabs.io                         | -                     | containerization | 3-Clause BSD  |
| Snap         | https://snapcraft.io                      | -                     | containerization | propriertary  |
| Spack        | https://spack.io                          | Gamblin et al, 2015   | package manager  | MIT or Apache |
| Vagrant      | https://www.vagrantup.com                 | -                     | virtual machine  | MIT           |
| yum          | http://yum.baseurl.org                    | -                     | package manager  |               |
| Zero Install | https://0install.net                      | -                     | package manager  | GNU LGPL 2.1+ | {#tbl:basic-info}


| Distribution System Name | Supported Operating Systems | Supported Languages | Root to Install | Root to Run |
| :----------- | :-------------------- | :---------| :------ | :------ |
| AppImage     | Linux                 | any       | n/a     | no      |
| APT          | Debian, Ubuntu        | any       | yes     | yes     |
| Bioconda     | Linux, macOS, Windows | any       | no      | no      |
| Bioconductor | Linux, macOS, Windows | R         | no      | no      |
| conda        | Linux, macOS, Windows | any       | no      | no      | 
| CRAN         | Linux, macOS, Windows | R         | no      | no      |
| Docker       | Linux, macOS, Windows | any       | yes     | no      |
| Easybuild    | Linux                 | any       | no      | no      |
| Flatpak      | Linux                 | any       | no      | no      |
| GNU Guix     | Linux                 | any       | no      | no      |
| Homebrew     | Linux, macOS          | any       | no      | no      |
| pip          | Linux, macOS, Windows | Python    | no      | no      |
| Singularity  | Linux, macOS          | any       | yes     | no      |
| Snap         | Linux                 | any       | yes     | no      |
| Spack        | Linux, macOS          | any       | no      | no      |
| Vagrant      | Linux, macOS, Windows | any       | yes     |         |
| yum          | Linux, macOS, Windows | any       | no      | yes     |
| Zero Install | Linux, macOS, Windows | any       | no      | no      | {#tbl:features} 


| Distribution System Name | First Release | Latest Release | Age in Years | Number of Releases | Number of Tools | Number of Bio Tools |
| :----------- | :--------- | :--------- | ------: | -------: | -------: | -------: |
| AppImage     | 2014-01-24 | 2020-06-01 | 7   | 121  | | |
| APT          | 1998-03-31 | 2020-05-08 | 22  | 362  | | |
| Bioconda     | 2014-01-24 | 2016-09-06 | 7   | 39   | | | 
| Bioconductor | 2002-05-01 | 2020-04-28 | 17  | 37   | | |
| conda        | 2014-01-24 | 2020-04-13 | 6   | 261  | | |
| CRAN         | 1997-04-23 | 2020-02-29 | 22  | 29   | | |
| Docker*      | 2013-03-23 | 2020-06-01 | 7   | 121  | | |
| Easybuild    | 2012-11-09 | 2020-04-14 | 7   | 51   | | |
| Flatpak      | 2015-03-23 | 2020-04-03 | 5   | 128  | | |
| GNU Guix     | 2012-07-07 | 2020-04-15 | 7   | 23   | | |
| Homebrew     | 2009-05-20 | 2020-05-04 | 10  | 155  | | |
| pip          | 2009-01-20 | 2020-04-28 | 11  | 81   | | |
| Singularity  | 2012-07-07 | 2020-04-15 | 7   | 23   | | |    
| snapd        | 2014-12-09 | 2020-07-15 | 5   | 232  | | |
| Spack        | 2014-07-09 | 2020-04-15 | 5   | 27   | | |
| Vagrant      |            |            |     |      | | | 
| yum**        | 2002-06-08 | 2011-06-28 | 18  | 221  | | |
| Zero Install | 2005-02-04 | 2020-05-04 | 15  | 145  | | | {#tbl:popularity}
*Docker Engine

**need to find someone with a redhat license who can confirm numbers 

| Distribution System Name | Official Repository Name | Repository URL | 
| :----------- | :-------------------------- | :------------------------------------------- |
| AppImage     | AppImageHub                 | https://appimage.github.io/apps/             |
| APT          | -                           | -                                            |
| Bioconda     | bioconda channel            | https://github.com/bioconda/bioconda-recipes |
| Bioconductor | -                           | https://www.bioconductor.org/packages/release/BiocViews.html#___Software |
| conda        | -                           | https://repo.anaconda.com/pkgs/              |
| CRAN         | -                           | https://cran.r-project.org/web/packages/available_packages_by_name.html |
| Docker       | Docker Hub                  | https://hub.docker.com                       |
| Easybuild    |                             |                                              |
| Flatpak      | Flathub                     | https://flathub.org                          |   
| GNU Guix     | -                           | https://guix.gnu.org/packages/               |
| Homebrew     | Homebrew Formulae           | https://formulae.brew.sh                     |
| pip          | Python Package Index (PyPI) | https://pypi.org                             |
| Singularity  | Singularity Hub             | https://singularity-hub.org                  |
| Snap         | Snapcraft                   | https://snapcraft.io/store                   |
| Spack        | -                           | -                                            |
| Vagrant      | Vagrant Cloud               | https://app.vagrantup.com/boxes/search       |
| yum          | -                           | -                                            |
| Zero Install | -                           | https://apps.0install.net                    | {#tbl:official-repositories} 
 

